File: VB_Code/ApplicationHub/ApplicationHub/PasswordGen.vb
'Harris Ransom
'Created 1/22/2019
Imports System

Public Class PasswordGen
#Region "Declaration statements"
    'Possible characters
    Dim lowercaseChars As String = "abcdefghijklmnopqrstuvwxyz"
    Dim uppercaseChars As String = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    Dim nums As String = "1234567890"
    Dim specialChars As String = "`~!@#$%^&*()-_=+;:{}[]\|,.<>?/'"

    'Password requirements
    Dim length As Integer = 0 'Length requirement
    Dim capitalReq As Integer = 0 'Uppercase requirement
    Dim numReq As Integer = 0 'Number requirement
    Dim specialReq As Integer = 0 'Special character requirement

    'Misc.
    Dim password As String = ""

    Dim lengthError As Boolean = False
    Dim capitalError As Boolean = False
    Dim numError As Boolean = False
    Dim specialError As Boolean = False
    Dim additionError As Boolean = False

    Dim randomLowerIndex As Integer = Math.Floor(Rnd() * 26)
    Dim randomUpperIndex As Integer = Math.Floor(Rnd() * 26)
    Dim randomNumIndex As Integer = Math.Floor(Rnd() * 10)
    Dim randomSpecialIndex As Integer = Math.Floor(Rnd() * 31)
#End Region

#Region "Methods"
    'Method to determine if there was an error
    Public Sub getError()
        'Length error
        If (length <= 0) Then
            lengthError = True
        End If
        'Capital error
        If (capitalReq < 0) Then
            capitalError = True
        End If
        'Number error
        If (numReq < 0) Then
            numError = True
        End If
        'Special error
        If (specialReq < 0) Then
            specialError = True
        End If
        'Addition error
        If ((specialReq + capitalReq + numReq) > length) Then
            additionError = True
        End If
    End Sub

    'Method to generate actual password
    Public Sub makePassword(requiredCount As Integer, characterSet As String)
        Dim localCount As Integer = 0

        'If the characterset is lowercase
        If (characterSet = "lower") And (Len(password) < length) Then
            While (localCount < requiredCount)
                password += lowercaseChars.Substring(randomLowerIndex, 1)
                localCount += 1
                randomLowerIndex = Math.Floor(Rnd() * 26)
            End While
        End If

        'If the characterset is uppercase
        If (characterSet = "upper") And (Len(password) <= length) Then
            While (localCount < requiredCount)
                password += uppercaseChars.Chars(randomUpperIndex)
                localCount += 1
                randomUpperIndex = Math.Floor(Rnd() * 26)
            End While
            localCount = 0
        End If

        'If the characterset is numbers
        If (characterSet = "nums") And (Len(password) <= length) Then
            While (localCount < requiredCount)
                password += nums.Chars(randomNumIndex)
                localCount += 1
                randomNumIndex = Math.Floor(Rnd() * 10)
            End While
            localCount = 0
        End If

        'If the characterset is special
        If (characterSet = "special") And (Len(password) <= length) Then
            While (localCount < requiredCount)
                password += specialChars.Chars(randomSpecialIndex)
                localCount += 1
                randomSpecialIndex = Math.Floor(Rnd() * 31)
            End While
            localCount = 0
        End If
    End Sub

    'Gets password
    Private Sub btnGetPass_Click(sender As Object, e As EventArgs) Handles btnGetPass.Click
        'Gets user input for password requirements
        length = Val(txtLength.Text)
        capitalReq = Val(txtCapital.Text)
        numReq = Val(txtNum.Text)
        specialReq = Val(txtSpecial.Text)

        'Gets any errors
        getError()

        'Makes password using methods, if there were no errors
        If (lengthError = False) And (capitalError = False) And (numError = False) And (specialError = False) And (additionError = False) Then
            makePassword(capitalReq, "upper") 'Uppercase
            makePassword(numReq, "nums") 'Numbers
            makePassword(specialReq, "special") 'Special characters
            makePassword(length - Len(password) + 1, "lower") 'Lowercase
        ElseIf (additionError) Then
            MessageBox.Show("Requirements do not add up!")
        ElseIf (lengthError) Then
            MessageBox.Show("Invalid length requirement!")
        ElseIf (capitalError) Then
            MessageBox.Show("Invalid capital requirement!")
        ElseIf (numError) Then
            MessageBox.Show("Invalid number requirement!")
        ElseIf (specialError) Then
            MessageBox.Show("Invalid special requirement!")
        End If

        lblResult.Text = "Password: " + password

        'Enables reset and copy buttons
        btnCopy.Enabled = True
        btnReset.Enabled = True
    End Sub

    'Reset Button 
    Private Sub btnReset_Click(sender As Object, e As EventArgs) Handles btnReset.Click
        txtCapital.Text = ""
        txtLength.Text = ""
        txtNum.Text = ""
        txtSpecial.Text = ""
        lblResult.Text = ""
        password = ""
        btnCopy.Enabled = False
        btnReset.Enabled = False
        Randomize()
        Me.Invalidate()
    End Sub

    'Copies password to clipboard
    Private Sub btnCopy_Click(sender As Object, e As EventArgs) Handles btnCopy.Click
        My.Computer.Clipboard.Clear()
        My.Computer.Clipboard.SetText(password)
    End Sub

    'Exits to the application hub
    Private Sub btnExit_Click(sender As Object, e As EventArgs) Handles btnExit.Click
        Application.Exit()
    End Sub
#End Region

#Region "Events"
    'Form load
    Private Sub PasswordGen_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Randomize()
    End Sub

    'KeyDown Event
    Private Sub PasswordGen_KeyDown(sender As Object, e As KeyEventArgs) Handles MyBase.KeyDown
        'If the escape key is pressed
        If (e.KeyCode = Keys.Escape) Then
            ApplicationHub.Show()
            Me.Close()
        End If
    End Sub
#End Region
End Class

File: Python_Code/UsefulFunctions/passwordGen.py
import sys
import secrets
import string
import random

#Defines all possible characters for the password
lowercase = string.ascii_lowercase
uppercase = string.ascii_uppercase
nums = string.digits
special_char = "!@#$%^&*()=+-~/[]{};:<>`"

#Defines global variables
password = ''
count = 0
length_error = False
capital_error = False
num_error = False
special_error = False

#Gets user input
input("Press ENTER to continue...")
length = int(input("Password Length?"))
capital_req = int(input("Capital letter requirement number?"))
num_req = int(input("Number requirement?"))
special_req = int(input("Special character requirement number?"))

#Creates methods
def error(length, capital, num, special):
    global length_error
    global capital_error
    global num_error
    global special_error
    if (length <= 0):
        length_error = True
    if (capital_req < 0):
        capital_error = True
    if (num_req < 0):
        num_error = True
    if (special_req < 0):
        special_error = True

def make_password(required_count, character_set):
    global count
    global password
    while (count < required_count):
        password += secrets.choice(character_set)
        count += 1
    count = 0

def shuffle(s):
    l = list(s)
    random.shuffle(l)
    return "".join(l)

#Displays error messages
error(length, capital_req, num_req, special_req)
if (length_error):
    print("Error! Invalid length!")
if (capital_error):
    print("Error! Invalid capital requirement!")
if (num_error):
    print("Error! Invalid number requirement!")
if (special_error):
    print("Error! Invalid special character requirement!")

#Determines if there was an error. If there wasn't, then the password is generated
if (not length_error) and (not capital_error) and (not num_error) and (not special_error):
    #Generates password
    make_password(capital_req, uppercase) #Makes uppercase
    make_password(num_req, nums) #Makes numbers
    make_password(special_req, special_char) #Majes special characters
    make_password(length - len(password), lowercase) #Makes lowercase
    #Displays the password
    print("PASSWORD: %s" % shuffle(password))
else:
    print("Please fix any errors.")



